### PR TITLE
Custom webpack config

### PR DESCRIPTION
**Pull request by cristianbote**
**What**:

Adding the possibility for packages to define a `frontity.config.*` file to customise the webpack, babel or frontity configurations.

**How**:

In order to be able to run custom configuration, we needed a way to import a new file named `frontity.config.(j|t)s` that exports configurations for each of the contexts: webpack, babel and frontity.

**Tasks**:

- [x] Code
- [x] TSDocs
- [x] TypeScript
- [x] Unit tests
- [x] End to end tests
- [x] TypeScript tests
- [x] Update community discussions
- [x] Add a changeset (with link to its [Feature Discussion](https://community.frontity.org/c/33) if it exists)

**Unrelated Tasks**

- [ ] Update starter themes
- [ ] Update other packages

**Additional Comments**

I've recorded a video about it here https://www.loom.com/share/a2da571c9dbc4858878fe9e4c73cfb4c

### CODE REVIEW

**Comment by luisherranz** [comment]
This is totally awesome. I'm pretty excited 😄

If you include the `frontity` configuration as well, as you showed in the video, we can create a package `@frontity/vercel` that uses the new `.vercel_build_output` folder for the build if `process.env.VERCEL` exists.

Another thing we want to add later to the `frontity.config.js` file is the ability to modify the settings and a priority. For example, the `@frontity/amp` package could set `mode: "amp"` automatically using this:

```js
// High priority, run before any other package.
export const priority = 1

// Set the mode to "amp" so the user doesn't have to.
export const settings = ({ settings }) => {
  settings.mode = "amp";
}
```

So definitely it's a good idea to use named exports 🙂👏

**Comment by cristianbote** [comment]
Hey folks! I believe this one is ready. So I've recored a summary for this PR to ease the review process https://www.loom.com/share/fd990be7bde14798a8348a403847f47f

**Comment by cristianbote** [comment]
@DAreRodz I've added @luisherranz as reviewer since you are on holiday 😃  enjoy it!

**Review by luisherranz — COMMENTED** [review]
Hey Cris, sorry I didn't answer this on Friday. I actually reviewed it, and I think it's great 🙂

I just have some minor comments which are shown below.

I don't want to approve this because I saw you activated the auto-merge and I was taking a look at the types and how to export types for the new `frontity.config.ts` functions:

```ts
import { WebpackCustomizer, BabelCustomizer } from "frontity/types";

export const webpack: WebpackCustomizer = ({ config }) => {
  // ..
};

export const babel: BabelCustomizer = ({ config }) => {
  // ..
};

```

I'd like to add that. After playing with types for a while last Friday, I think I have a proposal that I like, because it took a bit of type refactoring. 

Today I'm OOF, but tomorrow I'll finish it and push it to this branch so you can tell me what you think. And if you like it, we can include it as part of this PR.

File: .changeset/shiny-impalas-raise.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"@frontity/core": minor
+---
+
+Introducing the possbility to customize webpack with a configuration file.
```
[the remark this answers is not in the document]
Comment (cristianbote) [reply .changeset/shiny-impalas-raise.md]: Thank you! 😃  much, much better and explicit.

File: packages/core/src/scripts/utils/read-configuration.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+): Promise<Record<string, []>> => {
+  const packages = uniq(
+    sites.reduce((out, site) => {
+      return out.concat(site.packages);
+    }, [])
+  );
+  const dictionary = {};
+
+  for (const i in packages) {
+    const name = packages[i];
```
[the remark this answers is not in the document]
Comment (cristianbote) [reply packages/core/src/scripts/utils/read-configuration.ts:45]: I mean, _better_ is subjective 😄  the thing with `for..of` is that is getting transpiled as `for..in` is not. For example:

```js
// Does not get transpiled. What you write is what you get.
for (const i in list) {
  const entry = list[i];
}

// This one instead gets transpiled in...
for (const entry of list) {
}

// ...this
for (var _i = 0, _list = list; _i < _list.length; _i++) {
  var _entry = _list[_i];
}
```
In the end, you still get the `get` property accessor called, so, not really sure if is it better to use `for..in`? 🤷  I really have no preference. `for..in` was shorter to type.

Let me know what you think.

File: packages/core/src/scripts/utils/read-configuration.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    const hasConfig = await packageHasConfig(name);
+
+    if (hasConfig) {
+      const packageConfig = await import(createPath(name, "frontity.config"));
```
[the remark this answers is not in the document]
Comment (cristianbote) [reply packages/core/src/scripts/utils/read-configuration.ts]: Uh oh! That's a good catch 😄 ! Indeed, webpack needs the path in the string format, to be able to parse it.

**Comment by cristianbote** [comment]
Thank you so much @luisherranz! No worries with the types, anytime we can improve them. I'm still bad at them 😄  so.

**Review by luisherranz — APPROVED** [review]
I'm going to approve this so if you think the types refactoring (https://github.com/frontity/frontity/pull/812/commits/10a64b04c9482b731796c4ee3e12a63db56cb202) is fine, you can merge it.

**Comment by cristianbote** [comment]
Thank you @luisherranz!

I trust the types are more usable now. Which is amazing! I still feel I don't understand them sufficiently to weigh in though 😬 . Also the babel and frontity config tests are super awesome.

Thank you so much taking care of them 👍

**Pull request merged**